### ISSUE TITLE
Sequence Manager Implementation

### ISSUE THREAD

**Issue by joewandy** [glasgowcompbio/vimms#39]
We would like a flexible architecture that can cover a wide range of possibilities and experiments, e.g. DsDA style analysis, case-control study, etc:
1. Run controllers in a sequence
2. Update state based on run results
3. Fragment/not fragment precursor ions using the state information. This includes a list of interesting things to include/exclude for fragmentation. 
    - 'Interesting things' can be defined in many ways. 
    - For now, we consider the chromatographic peaks produced from peak picking, e.g. using XCMS, peakonly or mzMine2 (slow?).
Please discuss the requirement here before we proceeds to implementation.

For the next experiment, we aim to to test the following experiment on the machine:
1. Run fullscan (MS1) controller
2. Run peak picking to produce 'boxes', update boxes to state.
3. Target boxes in successive runs
    - We already have an [OptimalTopNController](https://github.com/sdrogers/vimms/blob/master/vimms/Controller/topN.py#L689-L795) that can do this.

@vinnydavies is there anything in https://github.com/sdrogers/vimms/blob/master/vimms/SequenceManager.py that can do the above? or needs to be completed?

**Comment by sdrogers** [comment]
Something for consideration here:
- I think we need the concept of a _state_
- The _state_ can be updated by passing it an individual scan, or multiple scans (a whole injection perhaps)
- The _state_ can be asked for its current prioritisation of things to fragment

Parts one and two are quite straightforward. Part 3 could be done in one of two ways:
1.  We send the state a scan and it tells us what to fragment
2. We ask it what should be fragmented (in general) and it tells us -- i.e. what it thinks is high priority now
3. We ask it what should be fragmented within some RT window (2 is like this, but for a very wide window)

This goes (I think) to the bottom of the choices to be made...

For the particular example mentioned...
- state starts empty
- peak picking happens at end of full scan and state is the set of boxes
- now we have a choice in how we implement things: either the scan is sent from the controller to the state, and the state says "fragment these" and updates its list of boxes or...
- the controller asks the state: "what are your priorities (i.e. the boxes)" it replies and then the controller sends the state back the list of scheduled scans so that it can update itself

Thoughts?

**Comment by sdrogers** [comment]
Maybe an alternative is that we pass all scans to the state to update itself.
And we then pass an MS1 scan, and it sends back a ranked list of the MS1 in the scan?

**Comment by joewandy** [comment]
Vinny is on holiday but he left a comment on slack, summarised below:

> For the experiment using optimal controller with peak picking in a loop, I wrote something that roughly did this before. In sequence manager [OptimalMultiSampleExperiment](https://github.com/sdrogers/vimms/blob/9f8f9f542e4bc29ce4c008feb091d756a27c380b/vimms/SequenceManager.py#L506-L517), no idea whether it still works. Depends whether optimal controller still works I guess.

> This does roughly what @joewandy's first comment on this issue asks. It just doesn't do it the way @simon suggested. The code is also in a serious need of reengineering. It made logical sense when I first wrote it, then became very clear soon after that my logic wasn't that solid. My number initial thoughts were to rearrange it with a separate section for processing, so this could be done on another core etc. 

> There is a half complete notebook called multi sample controllers in OneDrive. The first numbered section generally work and show roughly what the sequence manager can roughly do in terms of basic sequenced runs and gridsearch. There are some examples further down that run samples that are dependent on previous results, but these may not work any longer. I was in the process of updating these but hadnt quite got round to it yet

TODO: check the codes above and see how much can be reused, and what needs to be modified.

**Comment by joewandy** [comment]
Briefly discussed in the meeting earlier.

**Experimental manager (temporary name)**

Use case:
- Runs various controllers, stores states across controller runs
- Is given a mass spec object
- Is given an iterable sets of chemicals, i.e. sets of samples. 
    - *"Give me the next sample from a well, e.g. RA2 in the autosampler"*
- Passes sample (chemical set) to the mass specs, runs controllers in different configurations.
- It can also call peak picking.

Basically it's something that can define a *workflow*:
- Run something initially / after
- Run something if a condition is met
- Repeat running something until stopping condition

**Comment by joewandy** [comment]
Don't think this is needed anymore

**Comment by joewandy** [comment]
Not sure this is still needed, but sounds like something nice to have. Something to think about.

**Comment by joewandy** [comment]
State has now been implemented as 'grid' in Ross' work.
